### ISSUE TITLE
Bug in Nitsche's method

### ISSUE THREAD

**Issue by santiagobadia** [gridap/Gridap.jl#137]
Now, I have implemented Nitsche's method for Stokes.

Again, it does not work.

Let me know if you can see the bug, I thing I will go home, not able to see anything...

I am positive the error is in my part, but not sure where...

If you can find anything let me know

```julia
using Test
using Gridap
import Gridap: ∇
import LinearAlgebra: tr

##
const T = VectorValue{2,Float64}

u(x) = VectorValue(x[1]*(x[2]^2),0.0)
∇u(x) = TensorValue(x[2]^2,2.0*x[1]*(x[2]),0.0,0.0)
∇(::typeof(u)) = ∇u
Δu(x) = VectorValue(2.0*x[1],0.0)

p(x) = 0.0
∇p(x) = VectorValue(0.0,0.0)
∇(::typeof(p)) = ∇p

run_test(8,2)

function run_test(n,order)

  f(x) = -Δu(x) + ∇p(x)
  g(x) = tr(∇u(x)) # i.e. div(u(x))
  ud(x) = u(x)

  # Discrete model
  L = 1.0
  limits = (0.0, L, 0.0, L)
  ncellx = n
  model = CartesianDiscreteModel(domain=limits, partition=(ncellx,ncellx))

  γ = order*(order+1)
  γ0 = 1.0/10.0

  h = L / ncellx

  # Construct the FEspace 1
  fespace1 = FESpace( reffe=:Lagrangian, conformity=:H1, valuetype=VectorValue{2,Float64},
  model=model, order=order) #, diritags="boundary")

  # Construct the FEspace 2
  fespace2 = FESpace(
  reffe=:PLagrangian, conformity=:L2, valuetype=Float64,
  model=model, order=order-1, constraint=:zeromean)

  # Define test and trial
  Vh = TestFESpace(fespace1)
  Qh = TestFESpace(fespace2)
  Yh = [Vh, Qh]

  Uh = TrialFESpace(fespace1,ud)
  Ph = TrialFESpace(fespace2)
  Xh = [Uh, Ph]

  # Define integration mesh and quadrature for volume
  trian = Triangulation(model)
  quad = CellQuadrature(trian,degree=2*order)

  btrian = BoundaryTriangulation(model)
  bquad = CellQuadrature(btrian,degree=2*order)
  nb = NormalVector(btrian)

  # Weak form
  function A_Ω(y,x)
    u, p = x
    v, q = y
    inner(∇(v), ∇(u)) - inner(q,div(u)) - inner(div(v), p)
  end

  function B_Ω(y)
    v, q = y
    inner(v,f) - inner(q, g)
  end

  function A_∂Ω(y,x)
    u, p = x
    v, q = y
    (γ/h) * inner(v,u) - inner(v, ∇(u)*nb ) + inner(v, p*nb) - inner(∇(v)*nb, u) + inner(q*nb,u)
  end

  function B_∂Ω(y)
    v, q = y
    + (γ/h) * inner(v,ud) - inner(∇(v)*nb, ud) + inner(q*nb,ud)
  end

  t_Ω = AffineFETerm(A_Ω,B_Ω,trian,quad)

  t_∂Ω = AffineFETerm(A_∂Ω,B_∂Ω,btrian,bquad)

  op = LinearFEOperator(Yh,Xh,t_Ω,t_∂Ω)

  xh = solve(op)

  uh, ph = xh

  eu = u - uh

  ep = p - ph

  # writevtk(trian,"trian",cellfields=["uh"=>uh,"ph"=>ph, "eu"=>eu, "ep"=>ep])

  l2(u) = inner(u,u)
  h1(u) = inner(∇(u),∇(u)) + l2(u)

  eul2 = sqrt(sum( integrate(l2(eu),trian,quad) ))
  euh1 = sqrt(sum( integrate(h1(eu),trian,quad) ))
  epl2 = sqrt(sum( integrate(l2(ep),trian,quad) ))

  # @test eul2 < 1.e-8
  # @test euh1 < 1.e-8
  # @test epl2 < 1.e-8

  (eul2, euh1, epl2)
end
```

**Comment by fverdugo** [comment]
The usual suspects are the analytical functions and the terms in the weak form (specially the signs).

**Comment by santiagobadia** [comment]
I have simplified the test as much as possible. If the y component depends on the x component it does not work. I have solved now a vector Laplacian and copy-pasted the same terms as in your tutorial. The analytical solution is ridiculously simple.

If the x-component does not depend on the y-component (or vice versa) no problem.

```julia
u(x) = VectorValue(0.0, x[1])
∇u(x) = TensorValue(0.0,0.0,1.0,0.0)
∇(::typeof(u)) = ∇u
Δu(x) = VectorValue(0.0,0.0)

f(x) = - Δu(x)

ud(x) = u(x)

∇ud(x) = ∇u(x)

# Discrete model
L = 1.0
limits = (0.0, L, 0.0, L)
ncellx = 8
model = CartesianDiscreteModel(domain=limits, partition=(ncellx,ncellx))

h = L / ncellx

order = 2

γ = order*(order+1)
γ0 = 1.0/10.0

# Construct the FEspace 1
labels = FaceLabels(model)
fespace1 = FESpace( reffe=:Lagrangian, conformity=:H1, valuetype=VectorValue{2,Float64},
model=model, labels=labels, order=order )

# Define test and trial
Vh = TestFESpace(fespace1)
Uh = TrialFESpace(fespace1,ud)

# Define integration mesh and quadrature for volume
trian = Triangulation(model)
quad = CellQuadrature(trian,degree=2*order)

btrian = BoundaryTriangulation(model)
bquad = CellQuadrature(btrian,degree=2*order)
nb = NormalVector(btrian)

# Weak form
function A_Ω(v,u)
  inner(∇(v), ∇(u))
end

function B_Ω(v)
  inner(v,f)
end

function A_∂Ω(v,u)
  (γ/h) * inner(v,u) - inner(v, ∇(u)*nb ) - inner(∇(v)*nb, u)
end

function B_∂Ω(v)
    + (γ/h) * inner(v,ud)  - inner(∇(v)*nb, ud)
end

t_Ω = AffineFETerm(A_Ω,B_Ω,trian,quad)

t_∂Ω = AffineFETerm(A_∂Ω,B_∂Ω,btrian,bquad)

# Define the FEOperator
op = LinearFEOperator(Vh,Uh,t_Ω,t_∂Ω)

# Solve!
uh = solve(op)

# Define exact solution and error
eu = u - uh

writevtk(trian,"trian",cellfields=["uh"=>uh, "eu"=>eu])

# Define norms to measure the error
l2(u) = inner(u,u)
h1(u) = inner(∇(u),∇(u)) + l2(u)

eul2 = sqrt(sum( integrate(l2(eu),trian,quad) ))
euh1 = sqrt(sum( integrate(h1(eu),trian,quad) ))


@test eul2 < 1.e-8
@test euh1 < 1.e-8
```

**Comment by santiagobadia** [comment]
Exactly the same formulation for scalar Laplacian, it works.

Is `∇u*nb` correct? I.e., is this equal to `d_i u_j nb_j`?

I think this is the error... It should be `∇u*nb`

```julia
u(x) = x[1]
∇u(x) = VectorValue(1.0,0.0)
∇(::typeof(u)) = ∇u
Δu(x) = 0.0

f(x) = - Δu(x)

ud(x) = u(x)

∇ud(x) = ∇u(x)

# Discrete model
L = 1.0
limits = (0.0, L, 0.0, L)
ncellx = 8
model = CartesianDiscreteModel(domain=limits, partition=(ncellx,ncellx))

h = L / ncellx

order = 2

γ = order*(order+1)
γ0 = 1.0/10.0

# Construct the FEspace 1
labels = FaceLabels(model)
fespace1 = FESpace( reffe=:Lagrangian, conformity=:H1, valuetype=Float64,
model=model, labels=labels, order=order )

# Define test and trial
Vh = TestFESpace(fespace1)
Uh = TrialFESpace(fespace1,ud)

# Define integration mesh and quadrature for volume
trian = Triangulation(model)
quad = CellQuadrature(trian,degree=2*order)

btrian = BoundaryTriangulation(model)
bquad = CellQuadrature(btrian,degree=2*order)
nb = NormalVector(btrian)

# Weak form
function A_Ω(v,u)
  inner(∇(v), ∇(u))
end

function B_Ω(v)
  inner(v,f)
end

function A_∂Ω(v,u)
  (γ/h) * inner(v,u) - inner(v, ∇(u)*nb ) - inner(∇(v)*nb, u)
end

function B_∂Ω(v)
    + (γ/h) * inner(v,ud)  - inner(∇(v)*nb, ud)
end

t_Ω = AffineFETerm(A_Ω,B_Ω,trian,quad)

# t_∂Ω = FESource(B_∂Ω,btrian,bquad)
t_∂Ω = AffineFETerm(A_∂Ω,B_∂Ω,btrian,bquad)

# Define the FEOperator
op = LinearFEOperator(Vh,Uh,t_Ω,t_∂Ω)

# Solve!
xh = solve(op)
uh = xh

# Define exact solution and error
eu = u - uh

writevtk(trian,"trian",cellfields=["uh"=>uh, "eu"=>eu])

# Define norms to measure the error
l2(u) = inner(u,u)
h1(u) = inner(∇(u),∇(u)) + l2(u)

# Compute errors
eul2 = sqrt(sum( integrate(l2(eu),trian,quad) ))
euh1 = sqrt(sum( integrate(h1(eu),trian,quad) ))


@test eul2 < 1.e-8
@test euh1 < 1.e-8

##

end # module
```

**Comment by santiagobadia** [comment]
Hi @fverdugo 

```julia
 - inner(v, nb*∇(u) ) - inner(nb*∇(v), u)
```

does not work and I think it should work. 

It seems a bug of the `Fields` operations machinery, doesn't it?

Any idea how to circumvent this problem using the current machinery?

**Comment by fverdugo** [comment]
> Is ∇u*nb correct? I.e., is this equal to d_i u_j nb_j?
I think this is the error... It should be nb_i d_i u_j but I am not sure how to implement it.

Yes, `∇u*nb == d_i u_j nb_j`, and (in theory except a bug) `nb*∇u  == nb_i d_i u_j `.

I have found a work-around:

```julia
function A_∂Ω(v,u)
  (γ/h) * inner(v,u) - inner(outer(nb,v), ∇(u)) - inner(∇(v), outer(nb,u))
end

ud_cf = compose(ud,CellGeomap(btrian))

function B_∂Ω(v)
  + (γ/h) * inner(v,ud)  - inner(∇(v), outer(nb,ud_cf))
end

```

Here, you see that we have again the rigidity of `Function` vs `CellField`

**Comment by santiagobadia** [comment]
Ok, but this one is not only rigidity between `Function` and `CellField`. It is even for the test and trial functions. I think we should fix this problem because as it is now is cumbersome. And more expensive, you are using the double contraction of two tensors when it is not needed.

And if we fix it I would change the tutorials since the order above between the normal and gradient is the natural one that works in general (not only for scalar fields).

**Comment by fverdugo** [comment]
Yes, we have a bug in `nb*∇u` that we need definitively to solve. Doing this with the new machinery should be straight-forward.

**Comment by fverdugo** [comment]
These problems are fixed in v0.7.0. Closing issue.
